rust/src/lockfile.rs

```diff
@@ -0,0 +1,164 @@
+    let lf = lockfile_parse_stream(&mut f).map_err(|e| {
+        io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!("Parsing {}: {}", filename.to_string_lossy(), e.to_string()),
+        )
+    })?;
+    Ok(lf)
+}
+
+/*
+ * Format:
+ *
+ * {
+ *    "packages": [
+ *        [
+ *          "NEVRA1",
+ *          "<digest-algo>:<digest>"
+ *        ],
+ *        [
+ *          "NEVRA2",
+ *          "<digest-algo>:<digest>"
+ *        ],
+ *        ...
+ *    ]
+ * }
+ */
+#[derive(Serialize, Deserialize, Debug)]
+struct LockfileConfig {
+    // Core content
+    packages: Vec<(String, String)>,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    static VALID_PRELUDE_JS: &str = r###"
+{
+ "packages": [["foo", "repodata-foo"], ["bar", "repodata-bar"]]
+}
+"###;
+
+    static INVALID_PRELUDE_JS: &str = r###"
+{
+ "packages": [["foo", "repodata-foo"], ["bar", "repodata-bar"]],
+ "packages-x86_64": [["baz", "repodata-baz"]],
+ "packages-comment": "comment here",
+}
+"###;
+
+    #[test]
+    fn basic_valid() {
+        let mut input = io::BufReader::new(VALID_PRELUDE_JS.as_bytes());
+        let lockfile =
+            lockfile_parse_stream(&mut input).unwrap();
+        assert!(lockfile.packages.len() == 2);
+    }
+
+    #[test]
+    fn test_invalid() {
+        let mut input = io::BufReader::new(INVALID_PRELUDE_JS.as_bytes());
+        match lockfile_parse_stream(&mut input) {
+            Err(ref e) => match e.downcast_ref::<io::Error>() {
+                Some(ref ioe) if ioe.kind() == io::ErrorKind::InvalidInput => {}
+                _ => panic!("Expected invalid lockfile, not {}", e.to_string()),
+            },
+            Ok(_) => panic!("Expected invalid lockfile"),
+        }
+    }
+
+    #[test]
+    fn test_open_file_nonexistent() {
+        let path = "/usr/share/empty/manifest.json";
+        match lockfile_parse(path) {
+            Err(ref e) => assert!(e
+                .to_string()
+                .starts_with(format!("Can't open file {:?}:", path).as_str())),
+            Ok(_) => panic!("Expected nonexistent lockfile error for {}", path),
+        }
+    }
+}
+
+mod ffi {
+    use super::*;
+    use glib_sys;
+    use glib::translate::*;
+    use libc;
+    use std::ptr;
+
+    use crate::ffiutil::*;
+
+    #[no_mangle]
+    pub extern "C" fn ror_lockfile_get_map(
+        filename: *const libc::c_char,
+        gerror: *mut *mut glib_sys::GError,
+    ) -> *mut glib_sys::GHashTable {
+        // Convert arguments
+        let filename = ffi_view_os_str(filename);
+        match lockfile_parse(filename) {
+            Err(ref e) => {
+                error_to_glib(e, gerror);
+                ptr::null_mut()
+            },
+            Ok(lockfile) => {
+                let map = lockfile.packages
+                    .iter()
+                    .fold(HashMap::<String, String>::new(), |mut acc, (k, v)| {
+                        acc.insert(k.clone(), v.clone());
```

Comment:
I think you can use `.drain()` and avoid duping each string.

Reply:
And if we can use `BTreeMap` I'd be surprised if this whole thing wasn't just:
```
lockfile.packages.to_glib_full()
```

Reply:
Still don't know if it's a problem for `serde` but it looks like there is no `to_glib_full` for `BTreeMap` out of the box:
```
error[E0599]: no method named `to_glib_full` found for type `std::collections::BTreeMap<std::string::String, std::string::String>` in the current scope
   --> src/lockfile.rs:163:35
    |
163 |                 lockfile.packages.to_glib_full()
    |                                   ^^^^^^^^^^^^

```
I'll try to implement the trait myself :)